cuda/solver/batch_gmres_kernels.cu

```diff
@@ -76,40 +77,59 @@ namespace batch_gmres {
 template <typename T>
 using BatchGmresOptions = gko::kernels::batch_gmres::BatchGmresOptions<T>;
 
-#define BATCH_GMRES_KERNEL_LAUNCH(_stoppertype, _prectype)             \
-    apply_kernel<stop::_stoppertype<ValueType>>                        \
-        <<<nbatch, default_block_size, shared_size>>>(                 \
-            opts.max_its, opts.residual_tol, opts.restart_num, logger, \
-            _prectype<ValueType>(), a, bptr, xptr)
+#if GKO_CUDA_BATCH_GMRES_HAVE_NO_SHMEM
+
+#define BATCH_GMRES_KERNEL_LAUNCH(_stoppertype, _prectype)                    \
+    apply_kernel<stop::_stoppertype<ValueType>>                               \
+        <<<nbatch, default_block_size>>>(shared_gap, opts.max_its,            \
+                                         opts.residual_tol, opts.restart_num, \
+                                         logger, _prectype<ValueType>(), a,   \
+                                         bptr, xptr, workspace.get_data())
+#else
+
+
+#define BATCH_GMRES_KERNEL_LAUNCH(_stoppertype, _prectype)                 \
+    apply_kernel<stop::_stoppertype<ValueType>>                            \
+        <<<nbatch, default_block_size, shared_size>>>(                     \
+            shared_gap, opts.max_its, opts.residual_tol, opts.restart_num, \
+            logger, _prectype<ValueType>(), a, bptr, xptr)
+
+#endif
 
 template <typename BatchMatrixType, typename LogType, typename ValueType>
 static void apply_impl(std::shared_ptr<const CudaExecutor> exec,
                        const BatchGmresOptions<remove_complex<ValueType>> opts,
-                       LogType logger, const BatchMatrixType &a,
-                       const gko::batch_dense::UniformBatch<const ValueType> &b,
-                       const gko::batch_dense::UniformBatch<ValueType> &x)
+                       LogType logger, const BatchMatrixType& a,
+                       const gko::batch_dense::UniformBatch<const ValueType>& b,
+                       const gko::batch_dense::UniformBatch<ValueType>& x)
 {
     using real_type = gko::remove_complex<ValueType>;
     const size_type nbatch = a.num_batch;
-    const ValueType *const bptr = b.values;
-    ValueType *const xptr = x.values;
+    const ValueType* const bptr = b.values;
+    ValueType* const xptr = x.values;
 
     static_assert(default_block_size >= 2 * config::warp_size,
                   "Need at least two warps per block!");
 
     int shared_size =
-#if GKO_CUDA_BATCH_USE_DYNAMIC_SHARED_MEM
         gko::kernels::batch_gmres::local_memory_requirement<ValueType>(
             a.num_rows, b.num_rhs, opts.restart_num);
-#else
-        0;
-#endif
+    auto nrhs = b.num_rhs;
+    auto nrows = a.num_rows;
+    auto restart = opts.restart_num;
+    int shared_gap = 6 * nrows * nrhs + 3 * restart * nrhs +
```

Comment:
I guess a better name would be `global_gap`, to remain consistent with how it's used in BiCGStab. There, `shared_gap` means the gap between the beginnings to two vectors in shared memory in the same block.

Also, I should probably remove the inclusion of norms etc. in the local shared memory calculation, as we are now using static shared memory for those. Once that is done, you could just use `shared_size` for this.